File: src/desktop/modern/application/services/construct_tab/construct_tab_coordination_service.py
"""
Construct Tab Coordination Service

Handles coordination between construct tab components.
Replaces the large SignalCoordinator with focused coordination logic.
"""

from __future__ import annotations

from typing import Any

from PyQt6.QtCore import QObject, pyqtSignal

from desktop.modern.application.services.sequence.sequence_beat_operations import (
    SequenceBeatOperations,
)
from desktop.modern.application.services.sequence.sequence_start_position_manager import (
    SequenceStartPositionManager,
)
from desktop.modern.domain.models import BeatData, SequenceData


class ConstructTabCoordinationService(QObject):
    """
    Service for coordinating construct tab components.

    Responsibilities:
    - Component signal coordination
    - Sequence state management
    - UI transition coordination
    - Event handling and forwarding

    Replaces the 531-line SignalCoordinator with focused logic.
    """

    # External signals
    sequence_created = pyqtSignal(object)
    sequence_modified = pyqtSignal(object)
    start_position_set = pyqtSignal(str)
    generation_completed = pyqtSignal(bool, str)

    def __init__(
        self,
        beat_operations: SequenceBeatOperations,
        start_position_manager: SequenceStartPositionManager,
        layout_service,  # IConstructTabLayoutService
    ):
        super().__init__()
        self._beat_operations = beat_operations
        self._start_position_manager = start_position_manager
        self._layout_service = layout_service

        # Component references
        self._components: dict[str, Any] = {}

        # State tracking
        self._handling_sequence_modification = False

    def setup_component_coordination(self, components: dict[str, Any]) -> None:
        """Set up coordination between components."""
        self._components = components

        # Connect component signals
        self._connect_start_position_picker_signals()
        self._connect_option_picker_signals()
        self._connect_workbench_signals()
        self._connect_beat_operations_signals()
        self._connect_start_position_manager_signals()

    def _connect_start_position_picker_signals(self):
        """Connect start position picker signals."""
        start_position_picker = self._components.get("start_position_picker")
        if start_position_picker and hasattr(
            start_position_picker, "start_position_selected"
        ):
            start_position_picker.start_position_selected.connect(
                self._handle_start_position_selected
            )

    def _connect_option_picker_signals(self):
        """Connect option picker signals."""
        option_picker = self._components.get("option_picker")
        if option_picker and hasattr(option_picker, "pictograph_selected"):
            # Prevent duplicate connections
            try:
                option_picker.pictograph_selected.disconnect(
                    self._beat_operations.add_pictograph_to_sequence
                )
            except TypeError:
                pass  # Signal was not connected

            option_picker.pictograph_selected.connect(
                self._beat_operations.add_pictograph_to_sequence
            )

    def _connect_workbench_signals(self):
        """Connect workbench signals."""
        workbench = self._components.get("workbench")
        if workbench:
            if hasattr(workbench, "sequence_modified"):
                workbench.sequence_modified.connect(self._handle_workbench_modified)
            if hasattr(workbench, "operation_completed"):
                workbench.operation_completed.connect(self._handle_operation_completed)

    def _connect_beat_operations_signals(self):
        """Connect beat operations signals."""
        self._beat_operations.beat_added.connect(self._on_beat_modified)
        self._beat_operations.beat_removed.connect(self._on_beat_modified)
        self._beat_operations.beat_updated.connect(self._on_beat_modified)

    def _connect_start_position_manager_signals(self):
        """Connect start position manager signals."""
        self._start_position_manager.start_position_set.connect(
            self._on_start_position_set
        )
        self._start_position_manager.start_position_updated.connect(
            self._on_start_position_updated
        )

    def handle_sequence_modified(self, sequence: SequenceData) -> None:
        """Handle sequence modification events."""
        if self._handling_sequence_modification:
            return

        try:
            self._handling_sequence_modification = True
            self._update_ui_based_on_sequence(sequence)
            self.sequence_modified.emit(sequence)
        finally:
            self._handling_sequence_modification = False

    def handle_start_position_set(self, start_position: BeatData) -> None:
        """Handle start position set events."""
        self._start_position_manager.set_start_position(start_position)

    def handle_beat_added(self, beat_data: BeatData) -> None:
        """Handle beat added events."""
        self._beat_operations.add_pictograph_to_sequence(beat_data)

    def handle_generation_request(self, generation_config: dict[str, Any]) -> None:
        """Handle generation request events."""
        # For now, just emit completion signal
        # Actual generation logic would be implemented here
        self.generation_completed.emit(True, "Generation completed")

    def handle_ui_transition_request(self, target_panel: str) -> None:
        """Handle UI transition requests."""
        transition_map = {
            "start_position_picker": self._layout_service.transition_to_start_position_picker,
            "option_picker": self._layout_service.transition_to_option_picker,
            "graph_editor": self._layout_service.transition_to_graph_editor,
            "generate_controls": self._layout_service.transition_to_generate_controls,
            "export_panel": self._layout_service.transition_to_export_panel,
        }

        transition_method = transition_map.get(target_panel)
        if transition_method:
            transition_method()

    def _handle_start_position_selected(
        self, position_key: str, start_position_data: BeatData
    ):
        """Handle start position selection."""
        self._start_position_manager.set_start_position(start_position_data)
        self._layout_service.transition_to_option_picker()
        self.start_position_set.emit(position_key)

    def _handle_workbench_modified(self, sequence: SequenceData):
        """Handle workbench sequence modification."""
        self.handle_sequence_modified(sequence)

    def _handle_operation_completed(self, message: str):
        """Handle workbench operation completion."""
        # Could emit operation completed signal if needed
        pass

    def _on_beat_modified(self, *args):
        """Handle beat modification events."""
        # Get current sequence and update UI
        if len(args) >= 3:
            # beat_added signal with updated sequence
            updated_sequence = args[2]
            self.handle_sequence_modified(updated_sequence)
        else:
            # Other beat modification signals - get current sequence
            workbench = self._components.get("workbench")
            if workbench and hasattr(workbench, "get_sequence"):
                current_sequence = workbench.get_sequence()
                if current_sequence:
                    self.handle_sequence_modified(current_sequence)

    def _on_start_position_set(self, start_position_data: BeatData):
        """Handle start position set."""
        workbench = self._components.get("workbench")
        if workbench and hasattr(workbench, "set_start_position"):
            workbench.set_start_position(start_position_data)

    def _on_start_position_updated(self, start_position_data: BeatData):
        """Handle start position updated."""
        self._on_start_position_set(start_position_data)  # Same handling

    def _update_ui_based_on_sequence(self, sequence: SequenceData):
        """Update UI based on sequence state."""
        # Determine which panel to show based on sequence state
        has_start_position = self._start_position_manager.has_start_position()
        has_beats = sequence and sequence.length > 0 and not sequence.is_empty()

        if has_start_position or has_beats:
            self._layout_service.transition_to_option_picker()
        else:
            self._layout_service.transition_to_start_position_picker()

File: web/src/lib/services/implementations/construct/ConstructTabCoordinationService.ts
/**
 * Construct Tab Coordination Service - Implementation
 *
 * Coordinates between construct tab components (start position picker, option picker, workbench).
 * Based on desktop ConstructTabCoordinationService but simplified for web with runes.
 *
 * FIXED: Added proper state synchronization to resolve start position selection getting stuck
 */

import { GridMode } from "$lib/domain/enums";
import type {
  IConstructTabCoordinationService,
  IStartPositionService,
} from "../../interfaces/application-interfaces";
import type { BeatData, SequenceData } from "../../interfaces/domain-types";
import type {
  ISequenceService,
  IWorkbenchBeatOperationsService,
} from "../../interfaces/sequence-interfaces";

// Note: This service will need to be updated to use the new DI pattern
// For now, we'll comment out the direct import to avoid build errors
// import { sequenceStateService } from "../../SequenceStateService.svelte";

interface ComponentWithEventHandler {
  handleEvent?: (eventType: string, data: unknown) => void;
}

export class ConstructTabCoordinationService
  implements IConstructTabCoordinationService
{
  private components: Record<string, ComponentWithEventHandler> = {};
  private isHandlingSequenceModification = false;
  private eventListenersSetup = false;
  private boundEventHandlers: {
    startPositionSelected: (event: CustomEvent) => void;
    optionSelected: (event: CustomEvent) => void;
    sequenceModified: (event: CustomEvent) => void;
  } | null = null;

  constructor(
    private sequenceService: ISequenceService,
    private startPositionService: IStartPositionService,
    private workbenchBeatOperations: IWorkbenchBeatOperationsService
  ) {}

  /**
   * Clean up resources when service is destroyed
   */
  destroy(): void {
    this.disconnectComponentSignals();
    this.components = {};
  }

  setupComponentCoordination(
    components: Record<string, ComponentWithEventHandler>
  ): void {
    this.components = components;

    // Set up any cross-component communication here (only once)
    if (!this.eventListenersSetup) {
      this.connectComponentSignals();
      this.eventListenersSetup = true;
    }
  }

  async handleSequenceModified(sequence: SequenceData): Promise<void> {
    if (this.isHandlingSequenceModification) {
      return;
    }

    try {
      this.isHandlingSequenceModification = true;

      // **CRITICAL: Update singleton state to keep UI in sync**
      // TODO: Update to use new DI pattern
      // sequenceStateService.setCurrentSequence(sequence);

      // Update UI based on sequence state
      await this.updateUIBasedOnSequence(sequence);

      // Notify components about sequence change
      this.notifyComponents("sequence_modified", { sequence });
    } catch (error) {
      console.error("❌ Error handling sequence modification:", error);
    } finally {
      this.isHandlingSequenceModification = false;
    }
  }

  async handleStartPositionSet(startPosition: BeatData): Promise<void> {
    try {
      // Set loading state to show user something is happening
      // TODO: Update to use new DI pattern
      // sequenceStateService.setLoading(true);
      // sequenceStateService.clearError();

      // Set the start position using the service
      await this.startPositionService.setStartPosition(startPosition);

      // **CRITICAL: Create a sequence with the start position stored separately**
      // Create a new sequence with NO beats initially (progressive creation)
      const newSequence = await this.sequenceService.createSequence({
        name: `Sequence ${new Date().toLocaleTimeString()}`,
        length: 0, // Start with 0 beats - beats will be added progressively
        gridMode: GridMode.DIAMOND, // Default grid mode
        propType: "staff", // Default prop type
      });

      // **CRITICAL: Set the start position in the sequence's startPosition field, NOT as beat 0**
      const updatedSequence =
        await this.workbenchBeatOperations.setConstructionStartPosition(
          newSequence.id,
          startPosition
        );

      if (updatedSequence) {
        // **CRITICAL FIX: Update the singleton state that UI components watch**
        // TODO: Update to use new DI pattern
        // sequenceStateService.setCurrentSequence(updatedSequence);

        // Also clear loading state
        // sequenceStateService.setLoading(false);

        // Service completed successfully - let components handle additional state updates
        this.notifyComponents("sequenceCreated", {
          sequence: updatedSequence,
          startPosition: startPosition,
        });
      } else {
        console.error("❌ Failed to reload updated sequence");
        // TODO: Update to use new DI pattern
        // sequenceStateService.setLoading(false);
        // sequenceStateService.setError(
        //   "Failed to create sequence with start position"
        // );
        return;
      }

      // Notify components
      this.notifyComponents("start_position_set", { startPosition });

      // **CRITICAL: Notify components to update UI state**
      // The shouldShowStartPositionPicker logic should now automatically return false
      // because we've set a current sequence with a startPosition
      this.notifyComponents("ui_state_update_requested", {
        action: "hide_start_position_picker",
      });

      // Transition to option picker
      await this.handleUITransitionRequest("option_picker");
    } catch (error) {
      console.error("❌ Error handling start position set:", error);
      // TODO: Update to use new DI pattern
      // sequenceStateService.setLoading(false);
      // sequenceStateService.setError(
      //   `Failed to set start position: ${error instanceof Error ? error.message : "Unknown error"}`
      // );
    }
  }

  async handleBeatAdded(beatData: BeatData): Promise<void> {
    try {
      // Get current sequence from singleton state
      // TODO: Update to use new DI pattern
      const currentSequence = null; // sequenceStateService.currentSequence;

      if (!currentSequence) {
        console.error("❌ No current sequence available for adding beat");
        return;
      }

      // **CRITICAL: Use the workbench service to add the beat**
      // TODO: Fix when currentSequence is properly implemented
      // const updatedSequence = await this.workbenchBeatOperations.addBeat(
      //   currentSequence.id,
      //   beatData
      // );

      // **CRITICAL: Update singleton state**
      // TODO: Update to use new DI pattern
      // sequenceStateService.setCurrentSequence(updatedSequence);

      // Notify components
      this.notifyComponents("beat_added", { beatData });
    } catch (error) {
      console.error("❌ Error handling beat added:", error);
      // TODO: Update to use new DI pattern
      // sequenceStateService.setError(
      //   `Failed to add beat: ${error instanceof Error ? error.message : "Unknown error"}`
      // );
    }
  }

  async handleGenerationRequest(
    config: Record<string, unknown>
  ): Promise<void> {
    try {
      // TODO: Implement sequence generation
      // For now, just notify components
      this.notifyComponents("generation_requested", { config });

      // Simulate generation completion
      setTimeout(() => {
        this.notifyComponents("generation_completed", {
          success: true,
          message: "Generation completed",
        });
      }, 1000);
    } catch (error) {
      console.error("❌ Error handling generation request:", error);
    }
  }

  async handleUITransitionRequest(targetPanel: string): Promise<void> {
    try {
      // Emit custom events for UI transitions (similar to legacy implementation)
      const transitionEvent = new CustomEvent("construct-tab-transition", {
        detail: { targetPanel },
        bubbles: true,
      });

      if (typeof window !== "undefined") {
        document.dispatchEvent(transitionEvent);
      }

      // Notify components about the transition
      this.notifyComponents("ui_transition", { targetPanel });
    } catch (error) {
      console.error("❌ Error handling UI transition:", error);
    }
  }

  private connectComponentSignals(): void {
    // Set up event listeners for component coordination
    if (typeof window !== "undefined") {
      // Create bound event handlers to allow proper cleanup
      this.boundEventHandlers = {
        startPositionSelected: ((event: CustomEvent) => {
          this.handleStartPositionSet(event.detail.startPosition);
        }) as (event: CustomEvent) => void,

        optionSelected: ((event: CustomEvent) => {
          this.handleBeatAdded(event.detail.beatData);
        }) as (event: CustomEvent) => void,

        sequenceModified: ((event: CustomEvent) => {
          this.handleSequenceModified(event.detail.sequence);
        }) as (event: CustomEvent) => void,
      };

      // Add event listeners
      document.addEventListener(
        "start-position-selected",
        this.boundEventHandlers.startPositionSelected as EventListener
      );
      document.addEventListener(
        "option-selected",
        this.boundEventHandlers.optionSelected as EventListener
      );
      document.addEventListener(
        "sequence-modified",
        this.boundEventHandlers.sequenceModified as EventListener
      );
    }
  }

  /**
   * Clean up event listeners to prevent memory leaks and duplicate handlers
   */
  private disconnectComponentSignals(): void {
    if (typeof window !== "undefined" && this.boundEventHandlers) {
      document.removeEventListener(
        "start-position-selected",
        this.boundEventHandlers.startPositionSelected as EventListener
      );
      document.removeEventListener(
        "option-selected",
        this.boundEventHandlers.optionSelected as EventListener
      );
      document.removeEventListener(
        "sequence-modified",
        this.boundEventHandlers.sequenceModified as EventListener
      );
      this.boundEventHandlers = null;
      this.eventListenersSetup = false;
    }
  }

  private async updateUIBasedOnSequence(sequence: SequenceData): Promise<void> {
    try {
      // Determine which panel to show based on sequence state
      const hasStartPosition = sequence?.startingPositionBeat != null;
      const hasBeats = sequence && sequence.beats && sequence.beats.length > 0;

      let targetPanel: string;

      if (hasStartPosition || hasBeats) {
        targetPanel = "option_picker";
      } else {
        targetPanel = "start_position_picker";
      }

      // Transition to appropriate panel
      await this.handleUITransitionRequest(targetPanel);
    } catch (error) {
      console.error("❌ Error updating UI based on sequence:", error);
    }
  }

  private hasStartPosition(sequence: SequenceData): boolean {
    // **FIXED: Check the startPosition field instead of checking beats[0]**
    // This aligns with the modern architecture where start position is separate
    return sequence?.startingPositionBeat != null;
  }

  private notifyComponents(eventType: string, data: unknown): void {
    // Notify individual components if they have handlers
    Object.entries(this.components).forEach(([name, component]) => {
      if (component && typeof component.handleEvent === "function") {
        try {
          component.handleEvent(eventType, data);
        } catch (error) {
          console.error(`❌ Error notifying component ${name}:`, error);
        }
      }
    });

    // Emit global event for any listeners
    if (typeof window !== "undefined") {
      const event = new CustomEvent(`construct-coordination-${eventType}`, {
        detail: data,
        bubbles: true,
      });
      document.dispatchEvent(event);
    }
  }
}
